// @storybook/addon-actions/dist/preview/decorateAction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

exports.decorateAction = decorateAction;
exports.decorate = decorate;

var _action = require('./action');

var _action2 = _interopRequireDefault(_action);

var _actions = require('./actions');

var _actions2 = _interopRequireDefault(_actions);

var _withActions = require('./withActions');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function applyDecorators(decorators, actionCallback) {
  return function () {
    for (var _len = arguments.length, _args = Array(_len), _key = 0; _key < _len; _key++) {
      _args[_key] = arguments[_key];
    }

    var decorated = decorators.reduce(function (args, fn) {
      return fn(args);
    }, _args);
    actionCallback.apply(undefined, (0, _toConsumableArray3.default)(decorated));
  };
}

function decorateAction(decorators) {
  return function (name, options) {
    var callAction = (0, _action2.default)(name, options);
    return applyDecorators(decorators, callAction);
  };
}

function decorate(decorators) {
  var decorated = decorateAction(decorators);
  var decoratedActions = function decoratedActions() {
    var rawActions = _actions2.default.apply(undefined, arguments);
    var actionsObject = {};
    (0, _keys2.default)(rawActions).forEach(function (name) {
      actionsObject[name] = applyDecorators(decorators, rawActions[name]);
    });
    return actionsObject;
  };
  return {
    action: decorated,
    actions: decoratedActions,
    withActions: (0, _withActions.createDecorator)(decoratedActions)
  };
}
