// @storybook/addon-actions/dist/preview/withActions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDecorator = undefined;

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _entries = require('babel-runtime/core-js/object/entries');

var _entries2 = _interopRequireDefault(_entries);

var _global = require('global');

var _lodash = require('lodash.isequal');

var _lodash2 = _interopRequireDefault(_lodash);

var _addons = require('@storybook/addons');

var _addons2 = _interopRequireDefault(_addons);

var _coreEvents = require('@storybook/core-events');

var _coreEvents2 = _interopRequireDefault(_coreEvents);

var _actions = require('./actions');

var _actions2 = _interopRequireDefault(_actions);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var lastSubscription = void 0; // Based on http://backbonejs.org/docs/backbone.html#section-164

var lastArgs = void 0;

var delegateEventSplitter = /^(\S+)\s*(.*)$/;

var isIE = _global.Element != null && !_global.Element.prototype.matches;
var matchesMethod = isIE ? 'msMatchesSelector' : 'matches';

var root = _global.document && _global.document.getElementById('root');

var hasMatchInAncestry = function hasMatchInAncestry(element, selector) {
  if (element[matchesMethod](selector)) {
    return true;
  }
  var parent = element.parentElement;
  if (!parent) {
    return false;
  }
  return hasMatchInAncestry(parent, selector);
};

var createHandlers = function createHandlers(actionsFn) {
  for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }

  var actionsObject = actionsFn.apply(undefined, args);
  return (0, _entries2.default)(actionsObject).map(function (_ref) {
    var _ref2 = (0, _slicedToArray3.default)(_ref, 2),
        key = _ref2[0],
        action = _ref2[1];

    // eslint-disable-next-line no-unused-vars
    var _key$match = key.match(delegateEventSplitter),
        _key$match2 = (0, _slicedToArray3.default)(_key$match, 3),
        _ = _key$match2[0],
        eventName = _key$match2[1],
        selector = _key$match2[2];

    return {
      eventName: eventName,
      handler: function handler(e) {
        if (!selector || hasMatchInAncestry(e.target, selector)) {
          action(e);
        }
      }
    };
  });
};

var actionsSubscription = function actionsSubscription() {
  for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    args[_key2] = arguments[_key2];
  }

  if (!(0, _lodash2.default)(args, lastArgs)) {
    lastArgs = args;
    var handlers = createHandlers.apply(undefined, args);
    lastSubscription = function lastSubscription() {
      handlers.forEach(function (_ref3) {
        var eventName = _ref3.eventName,
            handler = _ref3.handler;
        return root.addEventListener(eventName, handler);
      });
      return function () {
        return handlers.forEach(function (_ref4) {
          var eventName = _ref4.eventName,
              handler = _ref4.handler;
          return root.removeEventListener(eventName, handler);
        });
      };
    };
  }
  return lastSubscription;
};

var createDecorator = exports.createDecorator = function createDecorator(actionsFn) {
  return function () {
    for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
      args[_key3] = arguments[_key3];
    }

    return function (story) {
      if (root != null) {
        _addons2.default.getChannel().emit(_coreEvents2.default.REGISTER_SUBSCRIPTION, actionsSubscription.apply(undefined, [actionsFn].concat(args)));
      }
      return story();
    };
  };
};

exports.default = createDecorator(_actions2.default);
